include/envoy/stats/stat_data_allocator.h

```diff
@@ -34,8 +35,8 @@ class StatDataAllocator {
    * @return CounterSharedPtr a counter, or nullptr if allocation failed, in which case
    *     tag_extracted_name and tags are not moved.
    */
-  virtual CounterSharedPtr makeCounter(absl::string_view name, std::string&& tag_extracted_name,
-                                       std::vector<Tag>&& tags) PURE;
+  virtual CounterSharedPtr makeCounter(absl::string_view name,
+                                       const TagProducer* tag_producer) PURE;
```

Comment:
It might make the code cleaner to have some type of `NullTagProducer` that just does nothing so you can pass by reference and avoid checks throughout the code, but up to you.